channeldb/payment_status.go

```diff
@@ -161,7 +184,131 @@ func (ps PaymentStatus) Registrable(settled, paymentFailed bool) error {
+// Based on the above variables, we derive the status using the following
+// table,
+//
+//	| inflight | settled | failed | reason |         status       |
+//	|:--------:|:-------:|:------:|:------:|:--------------------:|
+//	|   true   |   true  |  true  |   yes  |      StatusError     |
+//	|   true   |   true  |  true  |   no   |    StatusInFlight    |
+//	|   true   |   true  |  false |   yes  |      StatusError     |
+//	|   true   |   true  |  false |   no   |    StatusInFlight    |
+//	|   true   |   false |  true  |   yes  |    StatusInFlight    |
+//	|   true   |   false |  true  |   no   |    StatusInFlight    |
+//	|   true   |   false |  false |   yes  |    StatusInFlight    |
+//	|   true   |   false |  false |   no   |    StatusInFlight    |
+//	|   false  |   true  |  true  |   yes  |      StatusError     |
+//	|   false  |   true  |  true  |   no   |    StatusSucceeded   |
+//	|   false  |   true  |  false |   yes  |      StatusError     |
+//	|   false  |   true  |  false |   no   |    StatusSucceeded   |
+//	|   false  |   false |  true  |   yes  |      StatusFailed    |
+//	|   false  |   false |  true  |   no   |    StatusInFlight    |
+//	|   false  |   false |  false |   yes  |      StatusFailed    |
+//	|   false  |   false |  false |   no   |    StatusInitiated   |
+//
+// Overall, `settled` and `reason` cannot be both true otherwise initiating an
+// unexpected condition. Also, when `inflight`, `settled`, and `failed` are
+// false, this initiates the payment is newly created and hasn't made any HTLCs
+// yet.
+func decidePaymentStatus(htlcs []HTLCAttempt,
+	reason *FailureReason) PaymentStatus {
+
+	var (
+		inflight      bool
+		htlcSettled   bool
+		htlcFailed    bool
+		paymentFailed bool
+	)
+
+	// If we have a failure reason, the payment is failed.
+	if reason != nil {
+		paymentFailed = true
+	}
+
+	// Go through all HTLCs for this payment, check whether we have any
+	// settled HTLC, and any still in-flight.
+	for _, h := range htlcs {
+		if h.Failure != nil {
+			htlcFailed = true
+			continue
+		}
+
+		if h.Settle != nil {
+			htlcSettled = true
+			continue
+		}
+
+		// If any of the HTLCs are not failed nor settled, we
+		// still have inflight HTLCs.
+		inflight = true
+	}
+
+	// Use the DB state to determine the status of the payment.
+	switch {
+	// If any of the HTLCs is settled yet the payment is marked as failed,
+	// this initiates an error state.
+	//
+	// NOTE: whether we have failed or inflight HTLCs is irrelevant.
+	case htlcSettled && paymentFailed:
+		return StatusError
+
+	// If we have inflight HTLCs, no matter we have settled or failed
+	// HTLCs, or the payment failed, we still consider it inflight so we
+	// inform upper systems to wait for the results.
+	case inflight:
+		return StatusInFlight
+
+	// If we have no in-flight HTLCs, and at least one of the HTLCs is
+	// settled, the payment succeeded.
+	//
+	// NOTE: when reaching this case, paymentFailed must be false.
+	case htlcSettled:
```

Comment:
This one may also not match?

`|   false  |   true  |  false |   yes  |      StatusError     |`

Reply:
`| false | true | false | yes | StatusError |` would mean,
- we have no inflight htlcs
- we have at least one settled htlc
- we have no failed htlcs
- we have a failure reason for the payment

And this case won't end up here as it's been caught above.

Reply:
Same misinterpretation as https://github.com/lightningnetwork/lnd/pull/7175#discussion_r1098374780